File: src/Go/burn_test/garbage_allocator.go
package burn_test

import (
	. "../common"
	. "../common/nanotime"
	"time"
)

type GarbageAllocator struct {
	RunDuration   time.Duration
	Allocations   []AllocationInfo
	StartIndex    int32
	GarbageHolder *GarbageHolder

	// Statistics
	AllocationCount int64
	ByteCount       int64
	GCPauses        []Interval
}

type AllocationInfo struct {
	ArraySize       int32
	BucketIndex     int8
	GenerationIndex int8
}

func NewGarbageAllocator(runDuration time.Duration, allocations []AllocationInfo, startIndex int32) *GarbageAllocator {
	a := &GarbageAllocator{
		RunDuration:   runDuration,
		Allocations:   allocations,
		StartIndex:    startIndex,
		GarbageHolder: NewGarbageHolder(),
		GCPauses:      make([]Interval, 0, int32(runDuration.Seconds()*3000)), // No reallocations during the test
	}
	return a
}

func ByteSizeToArraySize(byteSize int32) int32 {
	s := (byteSize - MinAllocationSize + ArrayItemSize - 1) / ArrayItemSize
	if s < 0 {
		return 0
	}
	return s
}

func ArraySizeToByteSize(arraySize int32) int64 {
	return MinAllocationSize + (int64(arraySize) << ArrayItemSizeLog2)
}

func CreateGarbage(arraySize int32) []int64 {
	return make([]int64, arraySize)
}

func (a *GarbageAllocator) Run() {
	allocations := a.Allocations
	gh := a.GarbageHolder
	gcPauses := a.GCPauses

	allocationIndex := a.StartIndex
	allocationIndexMask := int32(len(allocations) - 1)

	a.GarbageHolder.Start()
	lastTimestamp := Nanotime().Nanoseconds()
	endTimestamp := lastTimestamp + a.RunDuration.Nanoseconds()

	for lastTimestamp < endTimestamp {
		ai := allocations[allocationIndex]
		allocationIndex = (allocationIndex + 1) & allocationIndexMask

		// Allocation
		var garbage = CreateGarbage(ai.ArraySize)
		gh.AddGarbage(garbage, ai.BucketIndex, ai.GenerationIndex)

		a.AllocationCount++
		a.ByteCount += ArraySizeToByteSize(ai.ArraySize)

		// Releasing what should be released by now
		if (allocationIndex & 15) == 0 {
			gh.TryReleaseGarbage()
		}

		// GC check
		elapsed := Nanotime().Nanoseconds()
		diff := elapsed - lastTimestamp
		if diff >= MinGCPause {
			gcPauses = append(gcPauses, Interval{
				Start: float64(lastTimestamp),
				End:   float64(elapsed),
			})
		}
		lastTimestamp = elapsed
	}
	a.GCPauses = gcPauses
}

File: src/CSharp/Main/BurnTest/GarbageAllocator.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Runtime.CompilerServices;
using Benchmarking.Common;

namespace GCBurn.BurnTest
{
    public class GarbageAllocator : IActivity
    {
        public const int ArrayItemSize = 8;
        public const int ArrayItemSizeLog2 = 3;
        public const int ObjectSize = 16; // Object header size on x64 platform
        public const int MinArraySize = ObjectSize + 8; // Object header + length
        public const int MinAllocationSize = MinArraySize + 8; // Array size + pointer size
        public const int AvgPauseFrequency = 12000;

        public TimeSpan RunDuration = TimeSpan.Zero;
        public long MinGCPause = Stopwatch.Frequency / 100_000; // 0.01 ms
        public (int ArraySize, sbyte BucketIndex, sbyte GenerationIndex)[] Allocations;
        public GarbageHolder GarbageHolder = new GarbageHolder();
        public int StartIndex;

        // Statistics
        public long AllocationCount;
        public long ByteCount;
        public List<Interval> GCPauses; 

        public GarbageAllocator(TimeSpan runDuration, (int, sbyte, sbyte)[] allocations, int startIndex)
        {
            RunDuration = runDuration;
            Allocations = allocations;
            StartIndex = startIndex;
            GCPauses = new List<Interval>((int) (runDuration.TotalSeconds * AvgPauseFrequency)); // No reallocations during the test
        }
        
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static int ArraySizeToByteSize(int arraySize) => MinAllocationSize + (arraySize << ArrayItemSizeLog2);

        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static int ByteSizeToArraySize(int byteSize) =>
            Math.Max(0, (byteSize - MinAllocationSize + ArrayItemSize - 1) / ArrayItemSize);
        
        [MethodImpl(MethodImplOptions.AggressiveInlining)]
        public static object CreateGarbage(int arraySize) => new long[arraySize];

        public void Run()
        {
            var duration = RunDuration;
            var allocations = Allocations;
            var gh = GarbageHolder;
            var gcPauses = GCPauses;

            var allocationIndex = StartIndex;
            int allocationIndexMask = Allocations.Length - 1; 

            GarbageHolder.Start();
            var lastTimestamp = Stopwatch.GetTimestamp();
            var endTimestamp = lastTimestamp + (long) (duration.TotalSeconds * Stopwatch.Frequency);

            while (lastTimestamp < endTimestamp) {
                var (arraySize, bucketIndex, generationIndex) = allocations[allocationIndex];
                allocationIndex = (allocationIndex + 1) & allocationIndexMask;
                
                // Allocation
                var garbage = CreateGarbage(arraySize);
                gh.AddGarbage(garbage, bucketIndex, generationIndex);
                AllocationCount++;
                ByteCount += ArraySizeToByteSize(arraySize);
                
                // Releasing what should be released by now
                if ((allocationIndex & 15) == 0)
                    gh.TryReleaseGarbage();

                // GC check
                var elapsed = Stopwatch.GetTimestamp();
                var diff = elapsed - lastTimestamp;
                if (diff >= MinGCPause)
                    gcPauses.Add(new Interval(lastTimestamp, elapsed));
                lastTimestamp = elapsed;
            }
        }
    }
}
